Implement acceptance criterias

ardevelop opened the pull request:
@danschultz PTAL

We split PR into commits that correspond to  implementing each acceptance criteria.
Demo page: https://ardevelop.github.io/pre-flexit-training

ardevelop commented:
Next plan:
- Optimize new state creation, we're going to clone only affected objects to reduce memory consumption.
- Add more intuitive controls for Send Forward and Send Backward actions (e.g. custom context menu)
- Smarter logic for changing z-Index to consider elements that are overlapped.
- Add ability to select and outline elements on canvas, to be able to delete them and move by keyboard.

danschultz commented on deploy.sh:
```diff
@@ -0,0 +1,12 @@
+git checkout dev
+git branch -D gh-pages
+git checkout -b gh-pages dev
+npm install react react-dom
```
assuming your `package.json` exists, you should just be able to run `npm install`

danschultz commented on src/utils/model/base_model.ts:
```diff
@@ -0,0 +1,69 @@
+export abstract class BaseModel implements IClonable<BaseModel> {
+  abstract clone(): BaseModel;
+
+  updateAll(predicate: (model: UpdatablePropeties) => UpdatablePropeties): UpdatablePropeties {
```
typescript supports [polymorphic this](https://github.com/Microsoft/TypeScript/pull/4910). so you can just do `updateAll(..): this`;

danschultz commented on src/utils/model/base_model.ts:
```diff
@@ -0,0 +1,69 @@
+    return cloned || this;
+  }
+
+  update<T extends UpdatablePropeties>(model: T, func: (model: T) => T): UpdatablePropeties {
```
typescript supports [polymorphic this](https://github.com/Microsoft/TypeScript/pull/4910). so you can just do `update(..): this`;

danschultz commented on src/redux/canvas.ts:
```diff
@@ -0,0 +1,81 @@
+  move(payload: ElementMoveAction) { this.call(MOVE, payload); }
+  sendForward(payload: ElementAction) { this.call(SEND_FORWARD, payload); }
+  sendBackward(payload: ElementAction) { this.call(SEND_BACKWARD, payload); }
+}
```
this is an interesting pattern that i haven't seen before in redux apps. it lgtm, but i'm curious how it scales as the app state gets more complicated.

danschultz commented on src/redux/canvas.ts:
```diff
@@ -0,0 +1,81 @@
+  sendBackward(payload: ElementAction) { this.call(SEND_BACKWARD, payload); }
+}
+
+export default CanvasDispatcher.default;
```
using this global throughout your app makes mocking and testing very difficult. i'd recommend instantiating instances for your dispatchers in the entry point of your app, and passing them down into your components.

danschultz commented:
👍  LGTM

ardevelop commented:
@danschultz PTAL
We've addressed your remarks, also added new functionality like selecting one or more (via Shift key) elements on canvas and ability to move them all.

danschultz commented on src/components/app/canvas/canvas.tsx:
```diff
@@ -0,0 +1,81 @@
+                );
+              })
+            }
+          </g>
```
consider putting this into it's own method and/or component.

danschultz commented on src/components/app/canvas/canvas.tsx:
```diff
@@ -0,0 +1,81 @@
+              );
+            }))
+          }
+        </g>
```
consider putting the selection into it's own method and/or component.

danschultz commented on src/components/app/canvas/element/element.tsx:
```diff
@@ -0,0 +1,115 @@
+    let selectedElementsOutlines: HTMLElement[];
+    let shiftKey = e.shiftKey;
+
+    new DragSession(
```
consider putting the drag behavior into a new selection component or maybe the canvas component. putting the logic here violates the principle that a component only modifies itself.

danschultz commented on src/app.ts:
```diff
@@ -0,0 +1,64 @@
+  get app(): App {
+    return (<ContextType>this.context).app;
+  }
+}
```
👍

danschultz commented on src/dispatchers/canvas.ts:
```diff
@@ -0,0 +1,137 @@
+  sendBackward(payload: ElementAction) { this.call(SEND_BACKWARD, payload); }
+  select(payload: ElementSelectAction) { this.call(SELECT, payload); }
+  selectFlush() { this.call(SELECT_FLUSH); }
+  selectMove(payload: MoveAction) { this.call(SELECT_MOVE, payload); }
```
Super minor, but I think the DX would be improved if you specified the args for each action, and the function created the action from those args.

danschultz commented:
@ardevelop added some additional comments. this PR is getting a bit large. could you merge this and submit a new one with additional changes?
